File: internal/hidraw/hidraw.go
// Package hidraw provides low-level HID transport via Linux hidraw ioctls.
// No cgo required — uses golang.org/x/sys/unix for syscalls.
package hidraw

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"unsafe"

	"golang.org/x/sys/unix"
)

// ioctl direction bits
const (
	iocWrite     = 1
	iocRead      = 2
	iocReadWrite = 3
)

func hidiocsfeature(length int) uintptr {
	return uintptr(iocReadWrite)<<30 | uintptr('H')<<8 | 0x06 | uintptr(length)<<16
}

func hidiocgfeature(length int) uintptr {
	return uintptr(iocReadWrite)<<30 | uintptr('H')<<8 | 0x07 | uintptr(length)<<16
}

// FindHidraw locates the /dev/hidrawN path for a given USB VID:PID.
func FindHidraw(vendorID, productID uint16) (string, error) {
	matches, err := filepath.Glob("/sys/class/hidraw/hidraw*/device/uevent")
	if err != nil {
		return "", err
	}
	sort.Strings(matches)

	for _, path := range matches {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		for _, line := range strings.Split(string(data), "\n") {
			if !strings.HasPrefix(line, "HID_ID=") {
				continue
			}
			// Format: HID_ID=BUS:VID:PID (each field variable-width hex)
			parts := strings.Split(strings.TrimPrefix(line, "HID_ID="), ":")
			if len(parts) < 3 {
				continue
			}
			vid, err1 := strconv.ParseUint(parts[1], 16, 32)
			pid, err2 := strconv.ParseUint(parts[2], 16, 32)
			if err1 != nil || err2 != nil {
				continue
			}
			if uint16(vid) == vendorID && uint16(pid) == productID {
				rel := strings.TrimPrefix(path, "/sys/class/hidraw/")
				name := strings.SplitN(rel, "/", 2)[0]
				return "/dev/" + name, nil
			}
		}
	}
	return "", fmt.Errorf("hidraw device %04x:%04x not found", vendorID, productID)
}

// HidrawDevice provides direct hidraw access for HID feature reports.
type HidrawDevice struct {
	Path string
	fd   int
}

// Open opens the hidraw device for reading and writing.
func (d *HidrawDevice) Open() error {
	fd, err := unix.Open(d.Path, unix.O_RDWR, 0)
	if err != nil {
		return fmt.Errorf("open %s: %w", d.Path, err)
	}
	d.fd = fd
	return nil
}

// Close closes the hidraw device.
func (d *HidrawDevice) Close() error {
	if d.fd > 0 {
		err := unix.Close(d.fd)
		d.fd = -1
		return err
	}
	return nil
}

// SendFeatureReport sends a SET_FEATURE report. buf[0] must be the report ID.
func (d *HidrawDevice) SendFeatureReport(buf []byte) error {
	_, _, errno := unix.Syscall(
		unix.SYS_IOCTL,
		uintptr(d.fd),
		hidiocsfeature(len(buf)),
		uintptr(unsafe.Pointer(&buf[0])),
	)
	if errno != 0 {
		return fmt.Errorf("HIDIOCSFEATURE: %w", errno)
	}
	return nil
}

// GetFeatureReport reads a GET_FEATURE report. reportID is placed in buf[0].
func (d *HidrawDevice) GetFeatureReport(reportID byte, length int) ([]byte, error) {
	buf := make([]byte, length)
	buf[0] = reportID
	_, _, errno := unix.Syscall(
		unix.SYS_IOCTL,
		uintptr(d.fd),
		hidiocgfeature(length),
		uintptr(unsafe.Pointer(&buf[0])),
	)
	if errno != 0 {
		return nil, fmt.Errorf("HIDIOCGFEATURE: %w", errno)
	}
	return buf, nil
}

File: avellcc/hidraw.py
"""Low-level hidraw interface for ITE 829x keyboard controllers."""

import os
import fcntl
import glob as _glob


def _IOC(direction, type_char, nr, size):
    return (direction << 30) | (ord(type_char) << 8) | nr | (size << 16)


def _HIDIOCSFEATURE(length):
    return _IOC(3, 'H', 0x06, length)


def _HIDIOCGFEATURE(length):
    return _IOC(3, 'H', 0x07, length)


# HIDIOCGRAWINFO = _IOC(2, 'H', 0x03, 8)
_HIDIOCGRAWINFO = 0x80084803


def find_hidraw(vendor_id: int, product_id: int) -> str | None:
    """Find the hidraw device path for a given USB VID:PID."""
    for path in sorted(_glob.glob('/sys/class/hidraw/hidraw*/device/uevent')):
        try:
            with open(path) as f:
                content = f.read()
            hid_id_line = [l for l in content.splitlines() if l.startswith('HID_ID=')]
            if not hid_id_line:
                continue
            parts = hid_id_line[0].split('=')[1].split(':')
            vid = int(parts[1], 16)
            pid = int(parts[2], 16)
            if vid == vendor_id and pid == product_id:
                hidraw_name = path.split('/sys/class/hidraw/')[1].split('/')[0]
                return f'/dev/{hidraw_name}'
        except (IOError, IndexError, ValueError):
            continue
    return None


class HidrawDevice:
    """Direct hidraw interface for sending/receiving HID feature reports."""

    def __init__(self, path: str):
        self.path = path
        self._fd = -1

    def open(self):
        self._fd = os.open(self.path, os.O_RDWR)

    def close(self):
        if self._fd >= 0:
            os.close(self._fd)
            self._fd = -1

    def __enter__(self):
        self.open()
        return self

    def __exit__(self, *args):
        self.close()

    def send_feature_report(self, data: bytes | bytearray | list[int]):
        """Send a SET_FEATURE report. First byte must be the report ID."""
        buf = bytearray(data)
        fcntl.ioctl(self._fd, _HIDIOCSFEATURE(len(buf)), buf)

    def get_feature_report(self, report_id: int, length: int) -> bytes:
        """Get a GET_FEATURE report. Returns `length` bytes including report ID."""
        buf = bytearray(length)
        buf[0] = report_id
        fcntl.ioctl(self._fd, _HIDIOCGFEATURE(length), buf)
        return bytes(buf)
